[CDAP-18966] Run kubeJobCleaner only in appfabric pod

kapadiap opened the pull request:
Currently the kubeJobCleaner background thread runs in both AppFabric and Preview services where only 1 is required.

Testing Done:
1. Change cdapmasters CR with a. trace log level for io.cdap.cdap.k8s.runtime b. reduce cleanup interval with program.container.cleaner.interval.mins: 10
2. Run pipeline with runtime param system.runtime.cleanup.disabled=true in k8s native env.
3. observe the appFabric and preview pod logs and confirm cleanup is only running in appfabric.
4. verify the pipeline specific entities are cleaned up after jobcleaner executes.

chtyim commented on cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java:
```diff
@@ -331,20 +332,22 @@ public void initialize(MasterEnvironmentContext context) throws IOException, Ill
         }
       }
     }
+    boolean enableKubeJobCleaner = podLabels.containsKey(APPFABRIC_POD_LABEL);
```
This is kind of ugly. Can we use the `MasterEnvironment.getTask()` and `MasterEnvironmentTask` instead? That is only executed in app-fabric.

kapadiap replied:
Updated the change to use MasterEnvironmentTask instead.

kapadiap committed:
[CDAP-18966] Run kubeJobCleaner only in appfabric pod

Currently the kubeJobCleaner background thread runs in both AppFabric and Preview services where only 1 is required.

Testing Done:

Change cdapmasters CR with a. trace log level for io.cdap.cdap.k8s.runtime. b. reduce cleanup interval with program.container.cleaner.interval.mins: 10
Run pipeline with runtime param system.runtime.cleanup.disabled=true in k8s native env.
observe the appFabric and preview pod logs and confirm cleanup is only running in appfabric.
verify the pipeline specific entities are cleaned up after jobcleaner executes.

chtyim commented on cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java:
```diff
@@ -353,6 +340,13 @@ public void onNamespaceDeletion(NamespaceDetail namespaceDetail) throws Exceptio
+  /**
+   * Returns the k8s label selector which selects all runs started by the k8s twill runner that has the run id label.
+   */
+  public String getK8sTwillResourcesSelector() {
```
This method name is repeating this class. This is already a k8s twill runner, so simply call this `getSelector()` is good enough.

chtyim commented on cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java:
```diff
@@ -153,8 +152,6 @@ public class KubeTwillRunnerService implements TwillRunnerService, NamespaceList
-  private final int jobCleanBatchSize;
   private final String selector;
   private final boolean enableMonitor;
   private final Gson gson;
```
I notice that this field is unused. Do we need it?

kapadiap replied:
removed this unused field

chtyim commented on cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java:
```diff
@@ -353,6 +340,13 @@ public void onNamespaceDeletion(NamespaceDetail namespaceDetail) throws Exceptio
+   * Returns the k8s label selector which selects all runs started by the k8s twill runner that has the run id label.
+   */
+  public String getK8sTwillResourcesSelector() {
+    return this.selector;
```
Remove `this.`

chtyim commented on cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeJobCleaner.java:
```diff
@@ -14,41 +14,47 @@
+  KubeJobCleaner(String selector, int batchSize, long delayMin) {
     this.selector = selector;
     this.batchSize = batchSize;
+    this.delayMillis = delayMin * 60 * 1000;
```
Use `TimeUnit.MINUTES.toMillis` instead.

chtyim commented on cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeJobCleaner.java:
```diff
@@ -82,6 +88,9 @@ public void run() {
         continuationToken = jobs.getMetadata().getContinue();
+      } catch (IOException e) {
+        LOG.warn("Failed to cleanup jobs. Error creating batch api client.", e);
+        break;
```
Should we try on this error as well? There could be temporary disconnection to the API server?

kapadiap replied:
hmm, I thought this was only looking at kubeconfig files cached locally, but it does reach out to the server as a last resort. We can retry on this error.

chtyim commented on cdap-master-spi/src/main/java/io/cdap/cdap/master/spi/environment/MasterEnvironment.java:
```diff
@@ -53,11 +54,11 @@ default void destroy() {
-   * Returns a {@link Optional} {@link MasterEnvironmentTask} to be executed periodically.
-   * It is guaranteed that there is no concurrent call to the task returned.
+   * Returns a list of {@link MasterEnvironmentTask}s to be executed periodically.
+   * It is guaranteed that there is no concurrent call to each task returned.
```
I don't think the current implementation is not guaranteeing no concurrent task execution?

kapadiap replied:
I am not too sure myself. The underlying AbstractSchedulerService provides a guarantee that the run iterations of a specific task will never run concurrently. However we can create multiple instances of the same task and schedule them all to run concurrently.

kapadiap replied:
Let me just remove the guarantee claim.

chtyim commented on cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java:
```diff
@@ -388,8 +389,8 @@ public Supplier<TwillRunnerService> getTwillRunnerSupplier() {
   @Override
-  public Optional<MasterEnvironmentTask> getTask() {
-    return Optional.ofNullable(podKillerTask);
+  public List<MasterEnvironmentTask> getTasks() {
```
I think using `Collection` is better as ordering is not important.

masoud-io commented on cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeJobCleaner.java:
```diff
@@ -14,41 +14,47 @@
+class KubeJobCleaner implements MasterEnvironmentTask {
   private static final Logger LOG = LoggerFactory.getLogger(KubeJobCleaner.class);
-  private final BatchV1Api batchV1Api;
+  private volatile BatchV1Api batchV1Api;
```
Add comments why this needs to be volatile

masoud-io reviewed (commented):
Please ensure that jobcleaner service does not run in system worker pods.

chtyim commented on cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java:
```diff
@@ -219,19 +218,16 @@ public class KubeMasterEnvironment implements MasterEnvironment {
-  public KubeMasterEnvironment() {
-    gson = new Gson();
-  }
+  public KubeMasterEnvironment() {}
```
We can remove this too since Java always generate a no-op default constructor if the class doesn't have a constructor.

chtyim commented on cdap-master/src/main/java/io/cdap/cdap/master/environment/k8s/AppFabricServiceMain.java:
```diff
@@ -164,8 +164,8 @@ protected void addServices(Injector injector, List<? super Service> services,
-    // Optionally adds the master environment task
-    masterEnv.getTask().ifPresent(task -> services.add(new MasterTaskExecutorService(task, masterEnvContext)));
+    // Adds the master environment tasks
+    masterEnv.getTasks().forEach(task -> services.add(new MasterTaskExecutorService(task, masterEnvContext)));
```
The MasterTaskExecutorService has a hardcode thread name. Since now we can run more than one master env task, it's better to have the thread name contains the task name as well. The `MasterEnvironmentTask` can have a getName method that default to the simple class name (`getClass().getSimpleName()`)

chtyim commented on cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java:
```diff
@@ -388,8 +384,8 @@ public Supplier<TwillRunnerService> getTwillRunnerSupplier() {
-  public Optional<MasterEnvironmentTask> getTask() {
-    return Optional.ofNullable(podKillerTask);
+  public Collection<MasterEnvironmentTask> getTasks() {
+    return SetUtils.hashSet(podKillerTask, kubeJobCleaner);
```
No need to use apache common lib in this case. Just return `Arrays.asList(podKillerTask, kubeJobCleaner)`.

chtyim commented on cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java:
```diff
@@ -342,14 +337,15 @@ public void initialize(MasterEnvironmentContext context) throws IOException, Ill
                                              workloadIdentityPool,
                                              workloadIdentityProvider);
+    kubeJobCleaner = new KubeJobCleaner(twillRunner.getSelector(),
+                                        Integer.parseInt(conf.getOrDefault(JOB_CLEANUP_BATCH_SIZE, "1000")),
```
Please define the default as private constant.

chtyim commented on cdap-master/src/main/java/io/cdap/cdap/master/environment/k8s/MasterTaskExecutorService.java:
```diff
@@ -47,7 +47,8 @@ final class MasterTaskExecutorService extends AbstractScheduledService {
   protected ScheduledExecutorService executor() {
-    executor = Executors.newSingleThreadScheduledExecutor(Threads.createDaemonThreadFactory("master-env-executor"));
+    executor = Executors.newSingleThreadScheduledExecutor(Threads.createDaemonThreadFactory(
+      "master-env-task-" + task.getName() + "-%d"));
```
Since this is a single thread executor, you don't need the `-%d` part.

kapadiap replied:
I see, I added it because the underlying ThreadFactory always expects such postfix integer format. I will remove this, the extra arg would be safely ignored.

kapadiap commented:
> Please ensure that jobcleaner service does not run in system worker pods.

Looked at the logs, the system worker pods are not running jobcleaner. The system worker pods use the MasterEnvironmentMain rather than the AppFabricServiceMain and hence do not schedule the kube cleaner task.

kapadiap committed:
Adds null check for podKillerTask

chtyim commented on cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java:
```diff
@@ -397,7 +397,12 @@ public Supplier<TwillRunnerService> getTwillRunnerSupplier() {
   @Override
   public Collection<MasterEnvironmentTask> getTasks() {
-    return Arrays.asList(podKillerTask, kubeJobCleaner);
+    Set<MasterEnvironmentTask> set = new HashSet<>(2);
```
Remove the `2`.

chtyim commented on cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java:
```diff
@@ -397,7 +397,12 @@ public Supplier<TwillRunnerService> getTwillRunnerSupplier() {
-    return Arrays.asList(podKillerTask, kubeJobCleaner);
+    Set<MasterEnvironmentTask> set = new HashSet<>(2);
+    set.add(kubeJobCleaner);
+    if (podKillerTask != null) {
```
Why is this needed? A task shouldn't be null?

kapadiap replied:
The podKillerTask is an optional and only initialized if some configurations are explicitly provided.
